discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -15,182 +15,448 @@
+        if documents.is_empty() {
+            return Ok(());
+        }
+
+        // The amount of documents that we can store via bulk inserts
+        // (https://docs.rs/sqlx-core/latest/sqlx_core/query_builder/struct.QueryBuilder.html#method.push_values)
+        // is limited by the sqlite bind limit.
+        // bind_limit divided by the number of fields in the largest tuple (NewsResource)
+        let documents = documents.iter().take(BIND_LIMIT / 9);
+
+        // Begin transaction
+        let mut tx = self
+            .pool
+            .begin()
+            .await
+            .map_err(|err| storage::Error::Database(err.into()))?;
+
+        // insert id into Document table (fk of HistoricDocument)
+        let mut query_builder = QueryBuilder::new("INSERT INTO Document (id) ");
+        query_builder.push_values(documents.clone(), |mut stm, doc| {
+            stm.push_bind(doc.id.as_uuid());
+        });
+        query_builder
+            .build()
+            .persistent(false)
+            .execute(&mut tx)
+            .await
+            .map_err(|err| storage::Error::Database(err.into()))?;
+
+        // insert id into HistoricDocument table
+        let mut query_builder = QueryBuilder::new("INSERT INTO HistoricDocument (documentId) ");
+        query_builder.push_values(documents.clone(), |mut stm, doc| {
+            stm.push_bind(doc.id.as_uuid());
+        });
+        query_builder
+            .build()
+            .persistent(false)
+            .execute(&mut tx)
+            .await
+            .map_err(|err| storage::Error::Database(err.into()))?;
+
+        // insert data into NewsResource table
+        let mut query_builder = QueryBuilder::new("INSERT INTO NewsResource (documentId, title, snippet, topic, url, image, datePublished, source, market) ");
+        query_builder.push_values(documents.clone(), |mut stm, doc| {
+            stm.push_bind(doc.id.as_uuid())
+                .push_bind(&doc.resource.title)
+                .push_bind(&doc.resource.snippet)
+                .push_bind(&doc.resource.topic)
+                .push_bind(doc.resource.url.to_string())
+                .push_bind(doc.resource.image.as_ref().map(ToString::to_string))
+                .push_bind(&doc.resource.date_published)
+                .push_bind(&doc.resource.source_domain)
+                .push_bind(format!("{}{}", doc.resource.language, doc.resource.country));
+        });
+        query_builder
+            .build()
+            .persistent(false)
+            .execute(&mut tx)
+            .await
+            .map_err(|err| storage::Error::Database(err.into()))?;
+
+        // insert data into NewscatcherData table
+        let mut query_builder =
+            QueryBuilder::new("INSERT INTO NewscatcherData (documentId, domainRank, score) ");
+        query_builder.push_values(documents.clone(), |mut stm, doc| {
+            stm.push_bind(doc.id.as_uuid())
+                .push_bind(doc.resource.rank as u32)
```

Comment:
in `document::NewsResource`the type of rank is `u64` should we cast here or should we change the type in the NewscatcherData table to be `BIGINT`?

Reply:
Does anyone have an opinion on this or should we leave it like that for now?

Reply:
right, INTEGER is i64 not u64 and BIGINT in sqlite is also i64 so even if you use BIGINT it _sliently_ case any number >=2**63 to a REAL which is fun when we fetch it... :unamused: 

So I guess for any u64 we have three choices:
- not use it
- bit cast to i64 when inserting and back to u64 when fetching
- proper cast to i64 if it fails either fail or clamp it

In this case rank is the domain rank, theoretically it's at most 30?bit because there are just that many domains but if it is over time, it could easily be many times the number of domain, but then 16 times is still only 35bit. Also if it's rank is very very bad or very very bad +1 doesn't really matter so I would go with a clamping cast from u64 to i64 for this case.

Reply:
so something like:
```rust
i64::try_from(u64_number).unwrap_or(i64::MAX);
u64::try_from(i64_number).unwrap_or(u64::MIN);
```
?